### DISCUSSION TITLE
Reports exported in DOC format does not show report content properly

### DISCUSSION THREAD

**Discussion by PenghaiZhang** [eclipse-birt/birt#1818] [General]
Hi everyone, I have report template created probably 10+ years ago. When my app integrated with BIRT 3.7.2, we could use this template to generate report in both different formats (e.g. HTML, PDF & DOC). Recently, we have upgrade BIRT to 4.16.0,  and then we found that the report in DOC format does not show contents properly. 

Basically, words in a `text` tag are shrunken to super narrow and become invisible. 

![image](https://github.com/user-attachments/assets/0fed117b-7340-4217-bb90-b3710b88c989)

![image](https://github.com/user-attachments/assets/de357b3a-18cc-40e1-809c-9b962c20aaa7)

You can see from the above screenshot where the top line and bottom line are broken.

Looking further to the table configuration, I saw the width is only 0.04cm.

![image](https://github.com/user-attachments/assets/7e414cc0-c013-4430-91dc-cbe57f6ae6d7)

We found that if we change the content type from HTML to auto or plain, the width will become normal again. However, we do want to support some HTML tags in our reports.

We also found that if we keep the content type as HTML but remove the text alignment, the content will be OK as well.

Can someone give me some ideas about this issue ??

**Comment by hvbtup** [comment]
What if you use BIRT to create a DOCX file instead? 
The proprietary DOC format is about to vanish sooner or later anyway. Many companies don't accept DOC files for security reasons.

**Reply by PenghaiZhang** [reply]
> What if you use BIRT to create a DOCX file instead? The proprietary DOC format is about to vanish sooner or later anyway. Many companies don't accept DOC files for security reasons.

Hi @hvbtup if we export the report in DOCX format, the shrunken contents will disappear.  If my above example, `<text id=417>` will disappear from the docx file. However, if I change the content type to auto or plain, the text will be visible in the report, but the HTML tags like `<h6>` and `<b>` will not work. So I am wondering maybe we should not use these HTML tags in non-html reports ?

**Reply by hvbtup** [reply]
What kind of "layout preference" do you use in your report?
For PDF, DOCX, DOX output you should use "Fixed Layout", because the page width is limited.
![grafik](https://github.com/user-attachments/assets/50196713-497c-40ba-bd4f-fe791d76a3dc)

BIRT supports a subset of HTML in dynamic text items and text items. This should work with most (if not all) emitters.
So, it is perfectly valid to use formatting tags like `<b>` or `<span style="...">`. I'm not sure about `<h6>`.

What font size are you specifying in your report design? You should preferably use absolute units like `pt` or `mm`, but not something like `px`.

**Reply by PenghaiZhang** [reply]
Hi @hvbtup ,

Yes, we use `fixed layout`.
![image](https://github.com/user-attachments/assets/c24e0142-0f8f-4457-925b-381f74852fe2)

In our reports,  in order to use HTML tags like `<b>` and `<h6>`, we need to use `html` as the content type. 
![image](https://github.com/user-attachments/assets/4449bb4f-6b2c-4864-b596-47a8aae77112)
But, the `textAlign` property then causes the trouble ( I found if I remove this property the report will be OK).

**Reply by PenghaiZhang** [reply]
And yes we use `pt` as font unit.

**Reply by PenghaiZhang** [reply]
The problem is our report used to work nicely with BIRT V3.7 but after we upgraded to V4.16 it fails to display report content properly.

**Reply by hvbtup** [reply]
So, it looks like a bug in BIRT to me at first glance.

Can you please perform a "binary search" and test your report with BIRT 4.8 - 4.16 to find out which was the first release to introduce this bug?

And can you please cerate a simple, self-contained example report (with no external code or data dependencies) that demonstrates the issue?

**Reply by PenghaiZhang** [reply]
[itemcount.zip](https://github.com/user-attachments/files/16384485/itemcount.zip)

Hi @hvbtup I have attached a report which should output one line text only. If you view this report in HTML or PDF it looks OK, but if you try doc format you will see the issue. 

I can confirm this issue exists in V4.9 and V4.16. Not so sure about V.4.10 - V4.15, but I image it should exist in these version. I will try to replicate the issue with V4.8.

Thank you so much!

**Reply by hvbtup** [reply]
Now that's what I call a minimal self-contained example!

I tested your report with an adapted version of BIRT 4.2.1 and in this version the DOC output still looks ok:
![grafik](https://github.com/user-attachments/assets/f1bf6144-631a-4c2c-b20f-b1d3ccbe9ece)

So, according to your tests, the issue was introduced somewhere between 4.2.1 and 4.9.

**Reply by hvbtup** [reply]
Hmm... With an adapted version of BIRT 4.12 and output in the DOCX format, the report still works.
The text is shown as Arial 24pt in my release of MS Word then, whereas the DOC output of BIRT 4.2.1 is Arial 20pt.
I also tested your report with an unmodified birt-report-designer-all-in-one-4.16.0-202406141054-win32.win32.x86_64.
The DOCX output looks the same as with the adapted version of BIRT 4.12.

So I cannot reproduce your issue.

However, I tested with DOCX format up to now.
So I also ran your report in the Web Viewer and exported it as "Word". This created the text in Arial 20pt.

How did you test DOC format? Certainly not from the All-In-One designer.

Please show your output of this report and how exactly you create it with BIRT.

**Reply by PenghaiZhang** [reply]
HI @hvbtup , I have created video to demo how I replicate the issue, but the video is too big to upload here so I sent an email to you directly. Can you please have a look ? Thank you!

**Reply by PenghaiZhang** [reply]
Ah man email is returned due to the video size. OK I will create a new one.

**Reply by PenghaiZhang** [reply]
https://github.com/user-attachments/assets/d98858aa-1e31-4994-a9e0-799813d635ad


https://github.com/user-attachments/assets/a45a8325-fa5d-4c08-a647-1b0fa238b61d

https://github.com/user-attachments/assets/a9bfce23-279a-402c-bf79-3fedfb1a6fb4

Hi @hvbtup , thanks for looking into our issue. I have attached 3 videos to demo how to replicate the format issue with doc and docx.

**Comment by hvbtup** [comment]
Maybe this is an issue with LibreOffice? Can you please attach the unmodified DOCX file that BIRT creates when you use "Run" / "View Report" / "As DOCX"  inside a ZIP here?

**Reply by PenghaiZhang** [reply]
[report.zip](https://github.com/user-attachments/files/16408792/report.zip)

Hi @hvbtup please checked the attached zip file.

**Comment by hvbtup** [comment]
This file looks perfectly OK when opened with MS Word:

![grafik](https://github.com/user-attachments/assets/0ba6399e-d682-4868-8126-30adc4eab0aa)

So this seems to be an issue with LibreOffice.

**Reply by PenghaiZhang** [reply]
Oh OK. I will verify it with MS Word tomorrow morning. 

What about the DOC format can I ask ?

**Comment by PenghaiZhang** [comment]
[broken docx.zip](https://github.com/user-attachments/files/16490140/broken.docx.zip)

![image](https://github.com/user-attachments/assets/eb31affd-a487-4635-9148-222165657a69)


Hi @hvbtup we have generated a DOCX file from the attached template and replicated the issue I mentioned above in MS Word. Can you please have a look ? BTW,  this DOCX file looks fine in LibreOffice.

**Reply by hvbtup** [reply]
The attached report doesn't work at all. It shows an error when opening. A data set is missing.

**Reply by PenghaiZhang** [reply]
Sorry @hvbtup  I didn't know that, but you can ignore that error and continue.

However, if you believe this error prevents further investigation, I will ask my support to fix this error first.
![image](https://github.com/user-attachments/assets/4a94f5f2-bede-4d42-849d-a12ad0c5ee0c)

**Reply by hvbtup** [reply]
Of course this error prevents investigation!

**Reply by PenghaiZhang** [reply]
I have requested my support to fix the error and I will attach the fixed report on next week. Thank you.

**Reply by PenghaiZhang** [reply]
Hi @hvbtup , our problem is the table is render with 1px width in MS Word, so we need to keep the table in the template. However, we do not really have a publicly available dataset. We have tried multiple times and confirmed that we can view the report with that missing dataset error. Hope you can give a try.

**Reply by hvbtup** [reply]
No. Please provide a self-contained report demonstrating the issue.

**Reply by PenghaiZhang** [reply]
[commercial_table_with_no_dataset.zip](https://github.com/user-attachments/files/16633474/commercial_table_with_no_dataset.zip)

@hvbtup The attached report is all good now. I do not see any error when using it in the designer. Can you give a go ?

**Reply by PenghaiZhang** [reply]
![image](https://github.com/user-attachments/assets/b81f4b0c-0c0b-4052-a12c-15c09144058a)

**Reply by PenghaiZhang** [reply]
> No. Please provide a self-contained report demonstrating the issue.

Hello @hvbtup we have provided a self-contained report. Can you please help investigate the format issue we reported ?

**Comment by hvbtup** [comment]
The report you provided contains (seemingly strange) size constraints.
It consists of a Grid containing a Table in a cell in the 4th row.
The Grid has a height of 2.344 in.
Thee first three rows have height of 1.177, 0.354, and 0.281 in.
This leaves 0.32 in for the last line which contains the table.
The cell with the table contains a label, then the table. The table contains a header line with labels in 10pt font size and a footer line with a label in 10pt font size, but in the footer, the label is nested in a 1x1 grid, so this has additional padding.
So the header and footer lines need at least 12pt and 14pt height, that gives 0.3611in. This alone is more than what is left for the last row of the grid, even without the detail rows. 
In the simplified example, we can ignore the detail row height due to the empty dataset. But still, the content cannot fit into the grid.

Furthermore, the master page is defined as 900mm x 210mm. This should not be a problem for PDF output, but I don't know if MS Word actually supports this width.

And you used the setting "auto layout"
![grafik](https://github.com/user-attachments/assets/c946d9bc-bfda-4e8b-8357-ce1d26cddc12)

I think this settings is only meant for HTML output.
AFAIK one should use "fixed layout" for PDF and Word output.
I changed this setting to "fixed layout".

OK, so the report is somewhat strange.

But we can reproduce the issue with it (using the DOCX emitter with BIRT 4.16).

The grid width is too high, because 34.948in + 2*0.5in (left and right page margin) gives more then 913mm, but the page width is only 900mm.
I deleted the grid width and height contraints since they don't make sense. This does not help.

In DOCX output, all the content is squashed into a very narrow column at the left, just like before:
![grafik](https://github.com/user-attachments/assets/2947b25c-7f46-4e8f-a9e8-9ee2fb745557)
If we click into this one-pixel column, we can see this better:
![grafik](https://github.com/user-attachments/assets/2042a3b3-cfea-4da4-b0b3-dde54f7c4077)

Also, the grid column has a size contraints.
Generally, I recommend not to use the mouse to adjust column sizes, because this adds size contraints to *all* columns *and* to the grid itself. This is not what you want usually. It's usually better to only specify exact sizes for some columns and leave one or more column widths empty to optimally use the available width.
I removed the explicit column width of the one and only grid column as well. This still did not solve the problem.

I also removed the styles and the whole inner table and still I can reproduce the issue.

In fact I can reproduce the issue with this radically simplified version for the report: Just a 1x1 grid:

```

<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.25" id="1">
    <property name="createdBy">Eclipse BIRT Designer Version 4.16.0.v202406141054</property>
    <property name="units">in</property>
    <property name="iconFile">/templates/blank_report.gif</property>
    <property name="layoutPreference">fixed layout</property>
    <property name="bidiLayoutOrientation">ltr</property>
    <property name="imageDPI">96</property>
    <page-setup>
        <simple-master-page name="Simple MasterPage" id="2">
            <property name="type">custom</property>
            <property name="orientation">landscape</property>
            <property name="height">210mm</property>
            <property name="width">900mm</property>
        </simple-master-page>
    </page-setup>
    <body>
        <grid id="7">
            <column id="8"/>
            <row id="346">
                <property name="height">0.3541666666666667in</property>
                <cell id="347">
                    <text id="348">
                        <property name="fontSize">14pt</property>
                        <property name="fontWeight">bold</property>
                        <property name="textAlign">left</property>
                        <property name="contentType">plain</property>
                        <text-property name="content"><![CDATA[Commercial Units by Qualification and Competencies]]></text-property>
                    </text>
                </cell>
            </row>
        </grid>
    </body>
</report>

```

**Reply by hvbtup** [reply]
I'll convert this to an issue because it obviously is one.

**Reply by speckyspooky** [reply]
The topic here is really the used master page and the conditions of Mircrosoft.
And Henning had the correct idea for the reason of all because for MS Word (and e.g. MS Excel too) the columns have a maximum width.
The maximum of a word column is "55.87 cm". If you have more a larger width or a width smaller like "0 cm" you will get an error message and the column won't be displayed correctly.
This is the reason to why we see the topic with 1 column earlier.
A workaround here woud be to set a column width e.g. "50 cm" and it will work - my example use exactly the maximum value.

![grafik](https://github.com/user-attachments/assets/935c069f-b9e8-48cc-9c4d-998388ce8366)

**Reply by speckyspooky** [reply]
Further details of the research result are given at the issue-ticket #1867

**Reply by PenghaiZhang** [reply]
Thank you guys so much for spending your time investigating this issue!!
